### DISCUSSION TITLE
How to get current page pathname in middleware

### DISCUSSION THREAD

**Discussion by mmdalipour** [vercel/next.js#37451] [Help]
Hey guys
as you guys know all requests including prefetched pages pass through middleware, I have some authentication logic that I only want to call whenever user redirects to a page not the ones that are prefetched

How can I access the current page pathname in middleware?

**Comment by icyJoseph** [comment]
```ts
import { NextResponse, type NextRequest } from "next/server";

export function middleware(req: NextRequest) {
  console.log(req.nextUrl.pathname, req.page)
  return NextResponse.next();
}
```

Though, it doesn't really tell you the pathname as a function of the file system routing structure. For example if a path is `/posts/[postId]`, then `req.nextUrl.pathname` sees, `/posts/some-unique-id`.

To get the **file system routing** name, use `name` from inside of `req.page`, that is, `req.page.name`, and also `req.page.params` for dynamic parameters.

**Reply by mmdalipour** [reply]
thank you joseph, sadly middleware also get calls for prefetched pages

let me make my question clear

assume we have a page called /test1 which has two next/link components that redirect user to /test2 and /test3
when we visit /test1 page next js will also prefetch /test2 and /test3 which then results in calling middleware for 3 times, one for /test1, and the others for /test2 and /test3, so req.nextUrl.pathname is pointing to all 3, i want to get the current page that browser is showing not the ones that are prefetched

so at the end i want to have something like this

```javascript
  if(req.nextUrl.pathname !== currentPagePathname) {
    console.log(`this page is prefetched`);
    return NextResponse.next();
  }

  // continue to authentication logic
```

**Reply by icyJoseph** [reply]
~~I guess, you could check the request header called 'referrer'? In the request object is directly called with `req.referrer`. Let me do some testing.~~

Mmm I can't really make it work with the referrer. Interesting use case, let me think a bit more.

### Edit

I had made a typo... `req.headers.get("Referer")` should do the trick, it gives you the source of the request, not only the pathname though.

**Reply by icyJoseph** [reply]
~~Alright, try with this:~~

~~Though I am worried about the need for a referer ('referrer' is the right word, but not the correct header name) header, and the assumption that when it is `null`, we are in the current page. Quite simply, sometimes the referer header is dropped for various reasons, but if it works in your use case, take it!~~

Failed attempt.

**Reply by mmdalipour** [reply]
yup, I tried referer and it point to previous url, i find that useful only when referer is empty for the current page (when we load app for the first time) but when we navigate through pages it's no longer empty nor for prefetched pages so i guess it can not help in this situation, or am i missing something?

**Reply by icyJoseph** [reply]
Right that's a problem as well, mmm let's, if we could catch the resolved url, we could see if it has **_next/data** or anything of that sort.

**Reply by icyJoseph** [reply]
It feels like there's no good way around this, unless there's some documentation we are missing. Perhaps you need to redesign your logic to always force users to be authenticated upon request, and that way you'll ensure that any further requests are also authenticated.

**Reply by mmdalipour** [reply]
thank you joseph for your time, you're right, maybe i need to redesign my logic, i tried all possible ways that i could think of but didn't get any answer

**Reply by icyJoseph** [reply]
Sorry, we couldn't crack it, perhaps someone else can, but after all of this, it looks like it might be a bit hard to maintain as well. 

Prefetched data through links, is fairly static. Prefetching means, it gets the JavaScript for the page, and if there's static generation JSON files, it gets those too. Of course, you may not want to make it public to just about anyone, but that's why any attempt at getting it should be rejected, if not properly authenticated.

So a page with getServerSideProps, is pre-fetched, as in the JavaScript that runs it, but not the getServerSideProps data, that is only fetched when you click on the given link.

**Reply by abdulhaseeb13mar** [reply]
export async function middleware(req: NextRequest) {
    console.log(
      `=======MIDDLEWARE RUNNING FOR '${req.url}'===================`
    );
 }`
 
 it does the trick
